src/scraper.py

```diff
@@ -0,0 +1,187 @@
+    return elem_is_tr and has_two_td
+
+
+def clean_data(element):
+    """
+    Takes a cell as an argument. Return the tag.string attribute
+    with extraneous characters stripped.
+    """
+    data = element.string
+    try:
+        return data.strip(" \n:-")
+    except AttributeError:
+        return u""
+
+
+def extract_restaurant_metadata(listing):
+    """
+    Take in the listing for a single restaurant.
+    Return a Python dict with metadata:
+    {<label>: <value>}.
+    """
+    listing_dict = {}
+    metadata_rows = listing.find('tbody').find_all(
+        has_two_tds, recursive=False
+    )
+    for row in metadata_rows:
+        for td in row.find_all('td', recursive=False):
+            key_row, val_row = row.find_all('td', recursive=False)
+            label_present = clean_data(key_row)
+            if label_present:
+                label_dict = label_present
+            else:
+                label_dict = ''
+            listing_dict.setdefault(label_dict, []).append(clean_data(val_row))
+    return listing_dict
+
+
+def is_inspection_row(element):
+    """A filter function. Take in an HTML element.
+    Return 'True' if the element matches the criteria:
+    1 - is a <tr> element,
+    2 - contains 4 <td> elements,
+    3 - word 'inspection' is in the text of it's first <td>,
+    4 - word 'inspection' is not the 1st word in that <td>.
+    Return 'False' otherwise.
+    """
+    is_tr = element.name == 'tr'
+    if not is_tr:
+        return False
+    td_children = element.find_all('td', recursive=False)
+    has_four_td = len(td_children) == 4
+    first_td_text = clean_data(td_children[0]).lower()
+    contains_word = 'inspection' in first_td_text
+    does_not_start = not first_td_text.startswith('inspection')
+    return is_tr and has_four_td and contains_word and does_not_start
+
+
+def extract_score_data(element):
+    """
+    Take a listing as an argument. Use a filter to find inspection rows.
+    Return a dict containing the average score, high score,
+    total inspection values.
+    """
+    inspection_rows = element.find_all(is_inspection_row)
+    samples = len(inspection_rows)
+    total = high_score = average = 0
+    for row in inspection_rows:
+        strval = clean_data(row.find_all('td')[2])
+        try:
+            intval = int(strval)
+        except (ValueError, TypeError):
+            samples -= 1
+        else:
+            total += intval
+            high_score = intval if intval > high_score else high_score
+    if samples:
+        average = total/float(samples)
+    data = {
+        u'Average Score': average,
+        u'High Score': high_score,
+        u'Total Inspections': samples
+    }
+    return data
+
+
+if __name__ == "__main__":
+    kwargs = {
+        'Inspection_Start': '1/1/2013',
+        'Inspection_End': '1/10/2013',
+        'City': 'Seattle',
+        'Zip_code': '98108'
+    }
+if len(sys.argv) > 1 and sys.argv[1] == 'test':
```

Comment:
These lines should be tabbed over, so they only execute when this script is called from the command line. Otherwise, these lines will run if this module is ever imported. Not that pertinent within the limited scope of this assignment, but as a general Python best practice, this is vital.